### PR TITLE
Added generic IsJuliaWrapper

### PR DESCRIPTION
**Pull request by sebasguts**
which can be used to wrap a Julia object and still be able to
call it on Julia functions

### CODE REVIEW

**Comment by sebasguts** [comment]
@fingolfin @ThomasBreuer Should we merge this?

**Comment by ThomasBreuer** [comment]
I would like to change JuliaExperimental to use this new feature,
thus it would be good to have it in the master branch.
Just the test failures are irritating.

**Comment by sebasguts** [comment]
Give me some minutes to fix that, please.

**Comment by sebasguts** [comment]
@ThomasBreuer On second thought, this is not super easy to fix atm, because we need to disable the automatic conversion for functions for this to work (i.e., we need to call `julia_gap` in the function handler).

I suggest we try to merge #160 today, and then this PR.

**Comment by ThomasBreuer** [comment]
@sebasguts Fine with me.

**Comment by fingolfin** [comment]
Clearly merging PR #160 "today" did no quite work out... And to fix JuliaExperimental in PR #160, we need this PR... so either we merge PR #160 ASAP, or else, the work in this PR is also added to PR #160, so that we can work on fixing JuliaExperimental there

File: JuliaInterface/gap/JuliaInterface.gd [hunk 1]
```diff
@@ -87,7 +87,20 @@ DeclareGlobalFunction( "JuliaIncludeFile" );
 #!  and then load these files with <Ref Func="JuliaIncludeFile"/>.
 DeclareGlobalFunction( "JuliaImportPackage" );
 
-DeclareCategory( "IsJuliaModule", IsObject );
+#! @Description
+#!  If any component or positional object is in this category,
```
Comment (fingolfin) [inline JuliaInterface/gap/JuliaInterface.gd]:
```suggestion
#!  If any component or positional object <C>obj</C> is in this category,
```

File: JuliaInterface/gap/JuliaInterface.gd [hunk 2]
```diff
@@ -87,7 +87,20 @@ DeclareGlobalFunction( "JuliaIncludeFile" );
...
+#!  a Julia function will not pass this object as an <C>MPtr</C> to
```
Comment (fingolfin) [inline JuliaInterface/gap/JuliaInterface.gd]:
```suggestion
#!  then calling a Julia function with <C>obj</C> as an argument will not pass it as an <C>MPtr</C>,
```

File: JuliaInterface/gap/JuliaInterface.gd [hunk 3]
```diff
@@ -87,7 +87,20 @@ DeclareGlobalFunction( "JuliaIncludeFile" );
...
+#!  Julia, but instead passes the Julia pointer stored in <C>JuliaPointer</C>.
```
Comment (fingolfin) [inline JuliaInterface/gap/JuliaInterface.gd]:
```suggestion
#!  but instead <C>JuliaPointer(obj)</C> is passed, which must be a Julia object.
```

File: JuliaInterface/gap/JuliaInterface.gd [hunk 4]
```diff
@@ -87,7 +87,20 @@ DeclareGlobalFunction( "JuliaIncludeFile" );
...
+#!  This allows wrapper for Julia pointers that behave like Julia
```
Comment (fingolfin) [inline JuliaInterface/gap/JuliaInterface.gd]:
```suggestion
#!  This allows implementing high-level wrapper objects for Julia pointers that behave like Julia
```

File: JuliaInterface/gap/JuliaInterface.gd [hunk 5]
```diff
@@ -87,7 +87,20 @@ DeclareGlobalFunction( "JuliaIncludeFile" );
...
+#!  objects for calls to Julia functions.
```
Comment (fingolfin) [inline JuliaInterface/gap/JuliaInterface.gd]:
```suggestion
#!  object when used as arguments in calls to Julia functions.
```

File: JuliaInterface/gap/JuliaInterface.gd [hunk 6]
```diff
@@ -87,7 +87,20 @@ DeclareGlobalFunction( "JuliaIncludeFile" );
...
+DeclareCategory( "IsTransparentJuliaWrapper", IsObject );
+
+#! @Description
+#!  Attribute for <C>IsTransparentJuliaWrapper</C>
```
Comment (fingolfin) [inline JuliaInterface/gap/JuliaInterface.gd]:
```suggestion
#!  Attribute for objects in the category <C>IsTransparentJuliaWrapper</C>.
```
Also, strictly speaking this should use a `<Ref>`

File: JuliaInterface/gap/JuliaInterface.gd [hunk 7]
```diff
@@ -87,7 +87,20 @@ DeclareGlobalFunction( "JuliaIncludeFile" );
...
+#!  Must be a Julia object.
```
Comment (fingolfin) [inline JuliaInterface/gap/JuliaInterface.gd]:
```suggestion
#!  The value of this object must be a Julia object.
```

File: JuliaInterface/src/convert.c [hunk 8]
```diff
@@ -25,6 +25,14 @@ jl_value_t * julia_gap(Obj obj)
     if (obj == False) {
         return jl_false;
     }
+    if (TNUM_OBJ(obj) == T_COMOBJ &&
+        CALL_1ARGS(JULIAINTERFACE_IsTransparentJuliaWrapper, obj) == True) {
+        Obj return_value = CALL_1ARGS(JULIAINTERFACE_JuliaPointer, obj);
+        if (!IS_JULIA_OBJ(return_value)) {
+            ErrorMayQuit("JuliaPointer is not a Julia object", 0, 0);
```
Comment (fingolfin) [inline JuliaInterface/src/convert.c]:
```suggestion
            ErrorMayQuit("JuliaPointer must be a Julia object (not a %s)", (int)TNAM_OBJ(return_value), 0);
```
That would fit better with the usual style of error messages (at least as used in the GAP kernel)

File: JuliaInterface/src/convert.c [hunk 9]
```diff
@@ -25,6 +25,14 @@ jl_value_t * julia_gap(Obj obj)
     if (obj == False) {
         return jl_false;
     }
+    if (TNUM_OBJ(obj) == T_COMOBJ &&
```
Comment (fingolfin) [inline JuliaInterface/src/convert.c]: The documentation claims that positional objects are supported, too, but here you only check for `T_COMOBJ`. So also `T_POSOBJ` should be here -- and for HPC-GAP also `T_APOSOBJ` and `T_ACOMOBJ`.

Comment (sebasguts) [reply JuliaInterface/src/convert.c]: Added those with the right `#ifdef` (hopefully). The formatting looks a bit weird.

**Comment by sebasguts** [comment]
I rebased the PR and adressed the comments.

File: JuliaInterface/src/convert.c [hunk 10]
```diff
@@ -29,6 +29,19 @@ jl_value_t * julia_gap(Obj obj)
...
+    UInt tnum = TNUM_OBJ(obj);
+    if ((tnum == T_COMOBJ || tnum == T_POSOBJ
+#ifdef HPCGAP
+         || tnum == T_APOSOBJ || tnum == T_ACOMOBJ
+#endif
+         ) &&
+        CALL_1ARGS(JULIAINTERFACE_IsTransparentJuliaWrapper, obj) == True) {
+        Obj return_value = CALL_1ARGS(JULIAINTERFACE_JuliaPointer, obj);
+        if (!IS_JULIA_OBJ(return_value)) {
+            ErrorMayQuit("JuliaPointer must be a Julia object (not a %s)", (int)TNAM_OBJ(return_value), 0);
```
[the remark this answers is not in the document]
Comment (sebasguts) [reply JuliaInterface/src/convert.c]: It probably isn't, because I just applied your suggestion ;)

**Comment by sebasguts** [comment]
@fingolfin I think I have adressed all your comments

File: JuliaInterface/gap/JuliaInterface.gd [hunk 11]
```diff
@@ -67,7 +70,22 @@ DeclareGlobalFunction( "JuliaIncludeFile" );
...
-DeclareCategory( "IsJuliaModule", IsObject );
+#! @Description
+#!  If any component or positional object <C>obj</C> is in this category,
+#!  then calling a Julia function with <C>obj</C> as an argument
+#!  will not pass it as an <C>MPtr</C>,
+#!  but instead <C>JuliaPointer(obj)</C> is passed, which must be a Julia object.
+#!  This allows implementing high-level wrapper objects
+#!  for Julia object that behave just like the wrapped Julia
+#!  objects when used as arguments in calls to Julia functions.
+DeclareCategory( "IsTransparentJuliaWrapper", IsObject );
```
[the remark this answers is not in the document]
Comment (sebasguts) [reply JuliaInterface/gap/JuliaInterface.gd]: Well, sure, but it probably is: "That is the first thing that came to my mind".
I will rename it :)

**Comment by sebasguts** [comment]
@fingolfin Yep, I am fine with the changes, thank you :)

**Pull request merged**